CoreFeatures.md

```diff
@@ -0,0 +1,178 @@
+the features available in a parquet-format release (and one of the
+implementations of it) and not on the *core feature list* are experimental.
+
+## Versioning
+
+This document is versioned by the parquet-format releases which follows the
+scheme of semantic versioning. It means that no feature will be deleted from
+this document under the same major version. (We might deprecate some, though.)
+Because of the semantic versioning if one implementation supports the core
+features of the parquet-format release `a.b.x` it must be able to read any
+parquet files written by implementations supporting the release `a.d.y` where
+`b >= d`.
+
+If a parquet file is written according to a released version of this document
+it might be a good idea to write this version into the field `compliance_level`
+in the thrift object `FileMetaData`.
+
+## Adding new features
+
+The idea is to only include features which are specified correctly and proven
+to be useful for everyone. Because of that we require to have at least two
+different implementations that are released and widely tested. We also require
+to implement interoperability tests for that feature to prove one
+implementation can read the data written by the other one and vice versa.
+
+## Core feature list
+
+This list is based on the [parquet thrift file](src/main/thrift/parquet.thrift)
+where all the data structures we might use in a parquet file are defined.
+
+### File structure
+
+All of the required fields in the structure (and sub-structures) of
+`FileMetaData` must be set according to the specification.
+The following page types are supported:
+* Data page V1 (see `DataPageHeader`)
+* Dictionary page (see `DictionaryPageHeader`)
+
+**TODO**: list optional fields that must be filled properly.
+
+### Types
+
+#### Primitive types
+
+The following [primitive types](README.md#types) are supported
+* `BOOLEAN`
+* `INT32`
+* `INT64`
+* `FLOAT`
+* `DOUBLE`
+* `BYTE\_ARRAY`
+* `FIXED\_LEN\_BYTE\_ARRAY`
+
+NOTE: The primitive type `INT96` is deprecated so it is intentionally not listed
+here.
+
+#### Logical types
+
+The [logical type](LogicalTypes.md)s are practically annotations helping to
+understand the related primitive type (or structure). Originally we have had
+the `ConvertedType` enum in the thrift file representing all the possible
+logical types. After a while we realized it is hard to extend and so introduced
+the `LogicalType` union. For backward compatibility reasons we allow to use the
+old `ConvertedType` values according to the specified rules but we expect that
+the logical types in the file schema are defined with `LogicalType` objects.
+
+The following LogicalTypes are supported:
+* `STRING`
+* `MAP`
+* `LIST`
+* `ENUM`
+* `DECIMAL` (for which primitives?)
+* `DATE`
+* `TIME`: **(Which unit, utc?)**
+* `TIMESTAMP`: **(Which unit, utc?)**
+* `INTEGER`: (all bitwidth 8, 16, 32, 64) **(unsigned?)**
+* `UNKNOWN` **(?)**
+* `JSON` **(?)**
+* `BSON` **(?)**
+* `UUID` **(?)**
```

Comment:
The last three seem highly application-specific. For example, Arrow doesn't have native types for them.

Reply:
This is a good point. I think this is true for logical types in general - if the application doesn't support the type at all, then it's a limitation of the application, not the application's Parquet implementation. It's reasonable for applications to not want to handle, say, timestamps.

Of course parquet-mr and parquet-cpp need to be able to at least pass through the data to the application in all of these cases.

So maybe all of these need to be qualified:
- If the implementation is general-purpose Parquet library, users of the library must be able to access data with all logical types. It may just return the data in the same format as the primitive type. E.g. if JSON is just returned as a STRING, that seems fine.
- If the implementation is anything else (SQL engine, computation framework, etc), then if it supports a primitive or logical type, it must support all core features associated with that type.

E.g. suppose you have an application that only processes DECIMAL values. It might be a bad and useless application, but that's not Parquet's problem. If it can read and write all DECIMAL encodings correctly, then it's compliant. If it couldn't read decimals stored as int64, or if it wrote decimals with precision < 10, it's not compliant (see https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal).

Reply:
That sounds fine to me.

Reply:
I think, this topic is not that simple. However parquet-mr usually returns values related to the primitive types only (no conversion for logical type) there are some additional logic that depends on the logical type.
* For example the sorting order matters when we are talking about statistics (min/max for row groups or column indexes). The sorting order might depend on the logical type (e.g. BINARY as a STRING or as a DECIMAL).
* In your example about JSON you cannot return a String if you don't know that the related BINARY primitive contains an UTF-8 encoded value.
* We have many problems with the different semantics used for timestamps in the different projects. This problem cannot be solved entirely in parquet but it can be the base of the solution. That's why we introduced the new timestamp logical type.
* Some parquet-mr bindings (e.g. parquet-avro) do depend on the logical types when converting values
* We already found some issues with the string representation. Impala did not write the logical type UTF-8 for strings only the primitive type BINARY. Meanwhile, iceberg required to have this logical type otherwise it recognized the value as a binary and it did not work with it.

To summarize, we might not add all of the logical types (or not all of the primitive representations) to the *core features* but I think, we shall add at least the ones already used widely. The whole *core features* idea is about to help interoperability. Logical types are a big part of it.

Although, from interoperability point of view a read-only parquet implementation that cannot do anything with the concept of DECIMAL it might not implement the reading of it. But if one would like to write a UTF-8 string than we shall require to use a STRING logical type in the schema for it.

Reply:
> In your example about JSON you cannot return a String if you don't know that the related BINARY primitive contains an UTF-8 encoded value.

In practice that shouldn't be very problematic, as most JSON should be UTF8-encoded (note that the [IETF standard](https://tools.ietf.org/html/rfc7159) requires either UTF-8, UTF-16 or UTF-32, and states that UTF-8 is the default encoding).

> The whole core features idea is about to help interoperability. Logical types are a big part of it.

Agreed. The question is: what does _supporting_ a logical type mean? For example, if I accept JSON data and simply present it as regular string data, is that enough to say that I support it? Otherwise, am I supposed to do something specific?

Reply:
My JSON example tried to point out that you have to recognize the logical type JSON to know it is a UTF-8 encoded string and not an arbitrary BINARY value. So in this case you at least partly support the logical type JSON.

As per the specification a value with the logical type JSON has to be a UTF-8 encoded json document (json spec referenced). What the parquet implementation does with it is not part of the parquet spec. So, from write point of view you  shall only use this annotation if it is really a json and is in UTF-8. From read point of view it is up to the implementation I guess.

I agree it is not always clear what to specify for supporting a logical type from reading. At least it shall respect the semantics of the related type. For example in case of TIMESTAMP if isAdjustedToUTC is true then the timestamp value shall be represented according to the *instant* semantics otherwise to the *local* semantics. (See details [here](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#timestamp).) Similarly even though UUID is represented in a binary format it is highly suggested to be represented in the UUID string format when displayed (if UUID logical type is supported).

Reply:
We shall agree on the list of the supported logical types and their primitive representations if there are more options. 

For the "support" point of view I would add something like the following.
> The writer shall support every listed logical types by either respecting its syntax/semantics or not using the related types at all. (For example if a writer does not use and therefore write any timestamp typed data it may still support the related version of core features.) The reader shall support every listed logical types where the related semantics matches one of its own data types. The reader can either convert the values of the other listed logical types to one of its own data type respecting the syntax/semantics (e.g. representing a timestamp as a string) or not reading these types. In the latter case the reader shall still recognize the related logical type and it shall be clear to the user that the file contains data that could not be represented. (The key here is if the reader has internal data representation for a semantics specified by one of these logical types it shall support reading it. Otherwise it shall either represent it in a meaningful way or not reading and notifying the user.)

What do you think?

Reply:
It's a bit difficult to read to me, what do you call "related type"?

Reply:
Yeah, I guessed so. I am not very good at it and that's why I've tried to write something here instead of in the document at the first place.

I would like to formalize the following idea. Writer writes a logical type if has a data type with similar semantics. It is allowed to not to write some logical types if there are no similar values to write (kind of common sense). Reader have to recognize all the listed logical types and either convert (or read as is) to one of its own types or if no such internal types exists it may not read such values but have to inform the user. We do not want misinterpreted or silently skipped values.

Reply:
So it is not ok for a reader to ignore the logical type and still interpret the data based on the physical type?

Reply:
It is not. Logical types can significantly modify the meaning of a primitive type. Back to one of my examples. A BINARY primitive type can be a STRING or a DECIMAL. Even the statistics (min/max) are different for these two.

Reply:
That makes sense, thank you.